File: trunk/Gomoku/Client/vb/Gomoku/Core/Game.vb
'
'  Gomoku Core
' 
'  Copyright (c) 2011 Tran Dinh Thoai <dthoai@yahoo.com>
'
' This library is free software; you can redistribute it and/or
' modify it under the terms of the GNU General Public License
' version 3.
'
' This library is distributed in the hope that it will be useful,
' but WITHOUT ANY WARRANTY; without even the implied warranty of
' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
' General Public License for more details.
'
' You should have received a copy of the GNU General Public License
' along with this library; if not, write to the Free Software
' Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
'

Option Explicit On
Option Strict On

Namespace com.bhivef.gomoku.core

    Public Class Game

        Public Const HUMAN_PLAYER As Byte = 0
        Public Const COMPUTER_PLAYER As Byte = 1
        Public Const REMOTE_PLAYER As Byte = 2

        Protected config As Config
        Protected board As Board
        Protected boardUI As BoardUI
        Protected client As Client
        Protected firstPlayer As Player
        Protected secondPlayer As Player
        Protected disposed As Boolean
        Protected finished As Boolean
        Private listeners As List(Of MoveListener)

        Public Sub New(ByVal config As Config, ByVal board As Board, ByVal boardUI As BoardUI, ByVal client As Client)
            Me.config = config
            Me.board = board
            Me.boardUI = boardUI
            Me.client = client
            Me.board.clear()
            Me.board.setCurrentPiece(core.Board.BLACK)
            Me.boardUI.clearListeners()
            Me.listeners = New List(Of MoveListener)
            Me.disposed = False
            Me.finished = False
        End Sub

        Public Sub create()
            If hasRemotePlayer() Then
                Dim playFirst As Boolean = (config.getByte(config.FIRST_TYPE) <> REMOTE_PLAYER)
                Dim gameId As String = client.createGame(playFirst, config.getInt(config.BOARD_WIDTH), config.getInt(config.BOARD_HEIGHT))
                config.setValue(config.CURRENT_GAME, gameId)
            End If
            createPlayers()
            setup()
            checkReady()
        End Sub

        Public Sub join(ByVal gameId As String)
            If hasRemotePlayer() Then
                client.joinGame(gameId)
                config.setValue(config.CURRENT_GAME, gameId)
            End If
            createPlayers()
            setup()
            checkReady()
        End Sub

        Public Sub list(ByVal tag As List(Of String))
            client.listGame(tag)
        End Sub

        Public Sub dispose()
            disposed = True
            listeners.Clear()
            If firstPlayer IsNot Nothing Then
                firstPlayer.dispose()
            End If
            If secondPlayer IsNot Nothing Then
                secondPlayer.dispose()
            End If
            Try
                client.endGame(False, board.NO_WIN)
            Catch
            End Try
        End Sub

        Private Sub setup()
            firstPlayer.addMoveListener(New MoveAdapter(Me, True))
            secondPlayer.addMoveListener(New MoveAdapter(Me, False))
        End Sub

        Protected Function hasRemotePlayer() As Boolean
            Dim remote As Boolean = False
            Dim firstType As Byte = config.getByte(config.FIRST_TYPE)
            Dim secondType As Byte = config.getByte(config.SECOND_TYPE)
            If firstType = REMOTE_PLAYER And secondType = REMOTE_PLAYER Then
                remote = False
            ElseIf firstType = REMOTE_PLAYER Or secondType = REMOTE_PLAYER Then
                remote = True
            End If
            Return remote
        End Function

        Protected Sub createPlayers()
            Dim remote As Boolean = False
            Dim firstType As Byte = config.getByte(config.FIRST_TYPE)
            Dim secondType As Byte = config.getByte(config.SECOND_TYPE)
            If firstType = REMOTE_PLAYER And secondType = REMOTE_PLAYER Then
                remote = False
            ElseIf firstType = REMOTE_PLAYER Or secondType = REMOTE_PLAYER Then
                remote = True
            End If

            If remote Then
                If firstType = REMOTE_PLAYER Then
                    firstPlayer = New RemotePlayer(config, client, board, board.BLACK)
                ElseIf firstType = COMPUTER_PLAYER Then
                    firstPlayer = New RemotePlayer(config, client, board, New ComputerPlayer(config, board, board.BLACK))
                ElseIf firstType = HUMAN_PLAYER Then
                    firstPlayer = New RemotePlayer(config, client, board, New HumanPlayer(config, board, board.BLACK, boardUI))
                Else
                    firstPlayer = createUnknownPlayer(True, firstType, board.BLACK)
                End If
                If secondType = REMOTE_PLAYER Then
                    secondPlayer = New RemotePlayer(config, client, board, board.WHITE)
                ElseIf secondType = COMPUTER_PLAYER Then
                    secondPlayer = New RemotePlayer(config, client, board, New ComputerPlayer(config, board, board.WHITE))
                ElseIf secondType = HUMAN_PLAYER Then
                    secondPlayer = New RemotePlayer(config, client, board, New HumanPlayer(config, board, board.WHITE, boardUI))
                Else
                    secondPlayer = createUnknownPlayer(True, secondType, board.WHITE)
                End If
            Else
                If firstType = COMPUTER_PLAYER Then
                    firstPlayer = New ComputerPlayer(config, board, board.BLACK)
                ElseIf firstType = HUMAN_PLAYER Then
                    firstPlayer = New HumanPlayer(config, board, board.BLACK, boardUI)
                Else
                    firstPlayer = createUnknownPlayer(False, firstType, board.BLACK)
                End If
                If secondType = COMPUTER_PLAYER Then
                    secondPlayer = New ComputerPlayer(config, board, board.WHITE)
                ElseIf secondType = HUMAN_PLAYER Then
                    secondPlayer = New HumanPlayer(config, board, board.WHITE, boardUI)
                Else
                    secondPlayer = createUnknownPlayer(False, secondType, board.WHITE)
                End If
            End If
        End Sub

        Protected Overridable Function createUnknownPlayer(ByVal hasRemote As Boolean, ByVal type As Byte, ByVal piece As Byte) As Player
            If hasRemote Then
                Return New RemotePlayer(config, client, board, New HumanPlayer(config, board, piece, boardUI))
            Else
                Return New HumanPlayer(config, board, piece, boardUI)
            End If
        End Function

        Public Sub addMoveListener(ByVal src As MoveListener)
            listeners.Add(src)
        End Sub

        Public Sub clearListeners()
            listeners.Clear()
        End Sub

        Protected Sub fireMoveMade(ByVal move As Move)
            For i As Integer = 0 To listeners.Count - 1
                listeners(i).moveMade(move)
            Next
        End Sub

        Protected Sub fireLastMove(ByVal victory As Byte)
            For i As Integer = 0 To listeners.Count - 1
                listeners(i).lastMove(victory)
            Next
        End Sub

        Protected Sub firstThink()
            Dim runner As New FirstThread(Me)
            Dim thread As New System.Threading.Thread(New System.Threading.ThreadStart(AddressOf runner.run))
            thread.Start()
        End Sub

        Protected Sub secondThink()
            Dim runner As New SecondThread(Me)
            Dim thread As New System.Threading.Thread(New System.Threading.ThreadStart(AddressOf runner.run))
            thread.Start()
        End Sub

        Protected Sub checkReady()
            Dim runner As New ReadyThread(Me)
            Dim thread As New System.Threading.Thread(New System.Threading.ThreadStart(AddressOf runner.run))
            thread.Start()
        End Sub

        Private Class FirstThread

            Private game As Game

            Public Sub New(ByVal game As Game)
                Me.game = game
            End Sub

            Public Sub run()
                game.firstPlayer.think(New Move())
            End Sub

        End Class

        Private Class SecondThread

            Private game As Game

            Public Sub New(ByVal game As Game)
                Me.game = game
            End Sub

            Public Sub run()
                game.secondPlayer.think(New Move())
            End Sub

        End Class

        Private Class ReadyThread

            Private game As Game

            Public Sub New(ByVal game As Game)
                Me.game = game
            End Sub

            Public Sub run()
                Dim stopIt As Boolean = False
                While Not stopIt And Not game.disposed
                    stopIt = game.firstPlayer.checkReady() And game.secondPlayer.checkReady()
                    If stopIt Then
                        Exit While
                    End If
                    Try
                        System.Threading.Thread.Sleep(1000)
                    Catch
                    End Try
                End While
                game.firstPlayer.setReady()
                game.secondPlayer.setReady()
                game.firstThink()
            End Sub

        End Class

        Private Class MoveAdapter
            Implements MoveListener

            Private game As Game
            Private first As Boolean

            Public Sub New(ByVal game As Game, ByVal first As Boolean)
                Me.game = game
                Me.first = first
            End Sub

            Public Sub lastMove(ByVal victory As Byte) Implements MoveListener.lastMove
                game.finished = True
                If first Then
                    game.secondPlayer.setFinished()
                Else
                    game.firstPlayer.setFinished()
                End If
                game.fireLastMove(victory)
                Try
                    game.client.endGame(True, victory)
                Catch
                End Try
            End Sub

            Public Sub moveMade(ByVal move As Move) Implements MoveListener.moveMade
                game.boardUI.update()
                game.fireMoveMade(move)
                If first Then
                    game.secondThink()
                Else
                    game.firstThink()
                End If
            End Sub

        End Class

    End Class

End Namespace

File: Gomoku/Client/cs/Gomoku/Core/Game.cs
/*
 *  Gomoku Core
 * 
 *  Copyright (c) 2011 Tran Dinh Thoai <dthoai@yahoo.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 3.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

using System;
using System.Collections.Generic;
using System.Text;

namespace com.bhivef.gomoku.core
{
    public class Game
    {
	    public const byte HUMAN_PLAYER = 0;
	    public const byte COMPUTER_PLAYER = 1;
	    public const byte REMOTE_PLAYER = 2;
	
	    protected Config config;
	    protected Board board;
	    protected BoardUI boardUI;
	    protected Client client;
	    protected Player firstPlayer;
	    protected Player secondPlayer;
	    protected bool disposed;
	    protected bool finished;
	    private List<MoveListener> listeners;

	    public Game(Config config, Board board, BoardUI boardUI, Client client) 
        {
		    this.config = config;
		    this.board = board;
		    this.boardUI = boardUI;
		    this.client = client;
		    this.board.clear();
		    this.board.setCurrentPiece(Board.BLACK);
		    this.boardUI.clearListeners();
		    this.listeners = new List<MoveListener>();
		    this.disposed = false;
		    this.finished = false;
	    }
	
	    public void create() 
        {
		    if (hasRemotePlayer()) 
            {
			    bool playFirst = (config.getByte(Config.FIRST_TYPE) != REMOTE_PLAYER);
			    String gameId = client.createGame(playFirst, config.getInt(Config.BOARD_WIDTH), config.getInt(Config.BOARD_HEIGHT));
			    config.setValue(Config.CURRENT_GAME, gameId);
		    }
		    createPlayers();
		    setup();
		    checkReady();
	    }
	
	    public void join(String gameId) 
        {
		    if (hasRemotePlayer()) 
            {
			    client.joinGame(gameId);
			    config.setValue(Config.CURRENT_GAME, gameId);
		    }
		    createPlayers();
		    setup();
		    checkReady();
	    }
	
	    public void list(List<String> tag) 
        {
		    client.listGame(tag);
	    }
	
	    public void dispose() 
        {
		    disposed = true;
		    listeners.Clear();
		    if (firstPlayer != null) 
            {
			    firstPlayer.dispose();
		    }
		    if (secondPlayer != null) 
            {
			    secondPlayer.dispose();
		    }
		    try 
            {
			    client.endGame(false, Board.NO_WIN);
		    } 
            catch { }
	    }
	
	    private void setup() 
        {
            firstPlayer.addMoveListener(new MoveAdapter(this, true));
            secondPlayer.addMoveListener(new MoveAdapter(this, false));
	    }
	
	    protected bool hasRemotePlayer() {
		    bool remote = false;
		    byte firstType = config.getByte(Config.FIRST_TYPE);
		    byte secondType = config.getByte(Config.SECOND_TYPE);
		    if (firstType == REMOTE_PLAYER && secondType == REMOTE_PLAYER) 
            {
			    remote = false;
		    }
            else if (firstType == REMOTE_PLAYER || secondType == REMOTE_PLAYER) 
            {
			    remote = true;
		    }
		    return remote;
	    }
	
	    protected void createPlayers() 
        {
		    bool remote = false;
		    byte firstType = config.getByte(Config.FIRST_TYPE);
		    byte secondType = config.getByte(Config.SECOND_TYPE);
		    if (firstType == REMOTE_PLAYER && secondType == REMOTE_PLAYER) 
            {
			    remote = false;
		    } 
            else if (firstType == REMOTE_PLAYER || secondType == REMOTE_PLAYER) 
            {
			    remote = true;
		    }
		    if (remote) 
            {
			    if (firstType == REMOTE_PLAYER) 
                {
				    firstPlayer = new RemotePlayer(config, client, board, Board.BLACK);
			    } 
                else if (firstType == COMPUTER_PLAYER) 
                {
				    firstPlayer = new RemotePlayer(config, client, board, new ComputerPlayer(config, board, Board.BLACK));
			    } 
                else if (firstType == HUMAN_PLAYER) 
                {
				    firstPlayer = new RemotePlayer(config, client, board, new HumanPlayer(config, board, Board.BLACK, boardUI));
			    } 
                else 
                {
				    firstPlayer = createUnknownPlayer(true, firstType, Board.BLACK);
			    }
			    if (secondType == REMOTE_PLAYER) 
                {
				    secondPlayer = new RemotePlayer(config, client, board, Board.WHITE);
			    } 
                else if (secondType == COMPUTER_PLAYER) 
                {
				    secondPlayer = new RemotePlayer(config, client, board, new ComputerPlayer(config, board, Board.WHITE));
			    } 
                else if (secondType == HUMAN_PLAYER) 
                {
				    secondPlayer = new RemotePlayer(config, client, board, new HumanPlayer(config, board, Board.WHITE, boardUI));
			    } 
                else 
                {
				    secondPlayer = createUnknownPlayer(true, secondType, Board.WHITE);
			    }
		    } 
            else 
            {
			    if (firstType == COMPUTER_PLAYER) 
                {
				    firstPlayer = new ComputerPlayer(config, board, Board.BLACK);
			    } 
                else if (firstType == HUMAN_PLAYER) 
                {
				    firstPlayer = new HumanPlayer(config, board, Board.BLACK, boardUI);
			    } 
                else 
                {
				    firstPlayer = createUnknownPlayer(false, firstType, Board.BLACK);
			    }
			    if (secondType == COMPUTER_PLAYER) 
                {
				    secondPlayer = new ComputerPlayer(config, board, Board.WHITE);
			    } 
                else if (secondType == HUMAN_PLAYER) 
                {
				    secondPlayer = new HumanPlayer(config, board, Board.WHITE, boardUI);
			    } 
                else 
                {
				    secondPlayer = createUnknownPlayer(false, secondType, Board.WHITE);
			    }
		    }
	    }
	
	    protected virtual Player createUnknownPlayer(bool hasRemote, byte type, byte piece) 
        {
		    if (hasRemote) 
            {
			    return new RemotePlayer(config, client, board, new HumanPlayer(config, board, piece, boardUI));
		    } 
            else 
            {
			    return new HumanPlayer(config, board, piece, boardUI);
		    }
	    }
	
	    public void addMoveListener(MoveListener src) 
        {
		    listeners.Add(src);
	    }
	
	    public void clearListeners() 
        {
		    listeners.Clear();
	    }
	
	    protected void fireMoveMade(Move move) 
        {
		    for (int i = 0; i < listeners.Count; i++) {
			    listeners[i].moveMade(move);
		    }
	    }
	
	    protected void fireLastMove(byte victory) 
        {
		    for (int i = 0; i < listeners.Count; i++) 
            {
			    listeners[i].lastMove(victory);
		    }
	    }
	
	    protected void firstThink() 
        {
            FirstThread runner = new FirstThread(this);
            System.Threading.Thread thread = new System.Threading.Thread(new System.Threading.ThreadStart(runner.run));
            thread.Start();
	    }

	    protected void secondThink() 
        {
            SecondThread runner = new SecondThread(this);
            System.Threading.Thread thread = new System.Threading.Thread(new System.Threading.ThreadStart(runner.run));
            thread.Start();
        }

	    protected void checkReady() {
            ReadyThread runner = new ReadyThread(this);
            System.Threading.Thread thread = new System.Threading.Thread(new System.Threading.ThreadStart(runner.run));
            thread.Start();
        }

        private class FirstThread
        {
            private Game game;

            public FirstThread(Game game)
            {
                this.game = game;
            }

            public void run()
            {
                game.firstPlayer.think(new Move());
            }
        }

        private class SecondThread
        {
            private Game game;

            public SecondThread(Game game)
            {
                this.game = game;
            }

            public void run()
            {
                game.secondPlayer.think(new Move());
            }
        }

        private class ReadyThread
        {
            private Game game;

            public ReadyThread(Game game)
            {
                this.game = game;
            }

            public void run()
            {
                bool stop = false;
                while (!stop && !game.disposed)
                {
                    stop = game.firstPlayer.checkReady() && game.secondPlayer.checkReady();
                    if (stop) break;
                    try
                    {
                        System.Threading.Thread.Sleep(1000);
                    }
                    catch { }
                }
                game.firstPlayer.setReady();
                game.secondPlayer.setReady();
                game.firstThink();
            }
        }

        private class MoveAdapter : MoveListener
        {
            private Game game;
            private bool first;

            public MoveAdapter(Game game, bool first)
            {
                this.game = game;
                this.first = first;
            }

            public void moveMade(Move move)
            {
                game.boardUI.update();
                game.fireMoveMade(move);
                if (first)
                {
                    game.secondThink();
                }
                else
                {
                    game.firstThink();
                }
            }

            public void lastMove(byte victory)
            {
                game.finished = true;
                if (first)
                {
                    game.secondPlayer.setFinished();
                }
                else
                {
                    game.firstPlayer.setFinished();
                }
                game.fireLastMove(victory);
                try
                {
                    game.client.endGame(true, victory);
                }
                catch { }
            }
        }

    }
}
